rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/topic_cache.hpp

```diff
@@ -102,81 +140,107 @@ class TopicCache
     return participant_to_topics_;
   }
 
+  /**
+   * @return a map of topic name to a vector of GUID_t, type name and qos profile tuple.
+   */
+  const TopicNameToTopicData & getTopicNameToTopicData() const
+  {
+    return topic_name_to_topic_data_;
+  }
+
   /**
    * Add a topic based on discovery.
    *
-   * @param rtpsParticipantKey
-   * @param topic_name
-   * @param type_name
+   * @param rtpsParticipantKey the participant key of the discovered publisher or subscription
+   * @param entity_guid the guid of the publisher or subscription
+   * @param topic_name the topic name associated with the discovered publisher or subscription
+   * @param type_name the topic type associated with the discovered publisher or subscription
+   * @param dds_qos the dds qos policy of the discovered publisher or subscription
    * @return true if a change has been recorded
    */
+  template<class T>
   bool addTopic(
     const eprosima::fastrtps::rtps::InstanceHandle_t & rtpsParticipantKey,
+    const GUID_t & entity_guid,
     const std::string & topic_name,
-    const std::string & type_name)
+    const std::string & type_name,
+    const T & dds_qos)
   {
-    initializeTopic(topic_name, topic_to_types_);
-    auto guid = iHandle2GUID(rtpsParticipantKey);
-    initializeParticipantMap(participant_to_topics_, guid);
-    initializeTopic(topic_name, participant_to_topics_[guid]);
+    initializeTopicDataMap(topic_name, topic_name_to_topic_data_);
+    auto participant_guid = iHandle2GUID(rtpsParticipantKey);
+    initializeParticipantMap(participant_to_topics_, participant_guid);
+    initializeTopicTypesMap(topic_name, participant_to_topics_[entity_guid]);
     if (rcutils_logging_logger_is_enabled_for("rmw_fastrtps_shared_cpp",
       RCUTILS_LOG_SEVERITY_DEBUG))
     {
       std::stringstream guid_stream;
-      guid_stream << guid;
+      guid_stream << entity_guid;
       RCUTILS_LOG_DEBUG_NAMED(
         "rmw_fastrtps_shared_cpp",
         "Adding topic '%s' with type '%s' for node '%s'",
         topic_name.c_str(), type_name.c_str(), guid_stream.str().c_str());
     }
-    topic_to_types_[topic_name].push_back(type_name);
-    participant_to_topics_[guid][topic_name].push_back(type_name);
+    rmw_qos_profile_t qos_profile = rmw_qos_profile_t();
+    dds_qos_to_rmw_qos(dds_qos, &qos_profile);
```

Comment:
Last detail that I forgot, this two lines are initing all the qos correctly, except `history kind` and `history depth`, which aren't available in the DDS discovery info (see discussion https://github.com/ros2/rmw_fastrtps/issues/319).
To avoid confusions, `history` should be set to [RMW_QOS_POLICY_HISTORY_UNKNOWN](https://github.com/ros2/rmw/blob/813b94ddd5650444b312304ad156f3b5d9f05e39/rmw/include/rmw/types.h#L274) and depth should be set to `0`. In that way, it's clear that the discovered value is unknown.

Reply:
I have added a `rmw_qos_profile_unknown` to `rmw/qos_profiles.h` that initializes `history_kind` and `history_depth` to what we want.

Reply:
Sounds great, thanks!